build.mill

```diff
@@ -0,0 +1,435 @@
+    "msg=method hasDefiniteSize in trait IterableOnceOps is deprecated:s", // replacement `knownSize` is not in 2.12
+    "msg=object JavaConverters in package collection is deprecated:s",
+    "msg=undefined in comment for method cf in class PrintableHelper:s",
+    // This is deprecated for external users but not internal use
+    "cat=deprecation&origin=firrtl\\.options\\.internal\\.WriteableCircuitAnnotation:s",
+    "cat=deprecation&origin=chisel3\\.util\\.experimental\\.BoringUtils.*:s",
+    "cat=deprecation&origin=chisel3\\.experimental\\.IntrinsicModule:s",
+    "cat=deprecation&origin=chisel3\\.ltl.*:s",
+    // Deprecated for external users, will eventually be removed.
+    "cat=deprecation&msg=Looking up Modules is deprecated:s",
+    // Only for testing of deprecated APIs
+    "cat=deprecation&msg=Use of @instantiable on user-defined types is deprecated:s"
+  )
+
+  // ScalacOptions
+  val scala2CommonOptions = Seq(
+    "-deprecation",
+    "-feature",
+    "-unchecked",
+    "-Werror",
+    "-Ymacro-annotations",
+    "-explaintypes",
+    "-Xcheckinit",
+    "-Xlint:infer-any",
+    "-Xlint:missing-interpolator",
+    "-language:reflectiveCalls",
+    s"-Wconf:${scala2WarnConf.mkString(",")}"
+  )
+}
+
+// def compileAll() = Task.Command {
+//   Task.traverse(v.buildUnits())(_.compile)()
+// }
+
+trait ChiselPublishModule extends CiReleaseModule {
+  // Publish information
+  def pomSettings = PomSettings(
+    description = artifactName(),
+    organization = "org.chipsalliance",
+    url = "https://www.chisel-lang.org",
+    licenses = Seq(License.`Apache-2.0`),
+    versionControl = VersionControl.github("chipsalliance", "chisel"),
+    developers = Seq(
+      Developer("jackkoenig", "Jack Koenig", "https://github.com/jackkoenig"),
+      Developer("azidar", "Adam Izraelevitz", "https://github.com/azidar"),
+      Developer("seldridge", "Schuyler Eldridge", "https://github.com/seldridge")
+    )
+  )
+
+  override def sonatypeHost = Some(SonatypeHost.s01)
+
+  override def publishVersion = VcsVersion
+    .vcsState()
+    .format(
+      countSep = "+",
+      revHashDigits = 8,
+      untaggedSuffix = "-SNAPSHOT"
+    )
+
+}
+
+trait HasScala2MacroAnno extends CrossModuleBase {
+  override def scalacOptions = Task {
+    if (!v.isScala3(crossScalaVersion)) {
+      super.scalacOptions() ++ Agg("-Ymacro-annotations")
+    } else super.scalacOptions()
+  }
+}
+
+trait HasScala2Plugin extends CrossModuleBase {
+  import $file.plugin.Plugin
+  def pluginModule: Plugin
+
+  override def scalacOptions = Task {
+    if (!v.isScala3(crossScalaVersion)) {
+      super.scalacOptions() ++ Agg(s"-Xplugin:${pluginModule.jar().path}")
+    } else super.scalacOptions()
+  }
+
+  override def scalacPluginClasspath = Task {
+    if (!v.isScala3(crossScalaVersion)) {
+      super.scalacPluginClasspath() ++ Agg(pluginModule.jar())
+    } else super.scalacPluginClasspath()
+  }
+}
+
+trait HasJextractGeneratedSources extends JavaModule {
```

Comment:
I think maybe we can reduce the size of `build.mill` be moving these functions to `package.mill`?

Reply:
Yes, we can, but they can't be accessed by `import build' this way, and will introduce a lot of `import $file`(or `import build_`) statements in `package.mill` of subfolders, which does not clarify the boundaries between modules and hepler files.